### PR TITLE
cell chargers now bypass APCs

### PR DESCRIPTION
**Pull request by Blacklist897**
## About The Pull Request
This makes cell chargers and suit storage units draw from the grid before the local apc

## Why It's Good For The Game
Upgraded Cell chargers have a charging power of 1MW leading to them instantly draining the apc of any room they are in,
this Pr makes them draw from the grid preventing immediate blackout. 
 This is a stopgap until someone smarter than me changes power values so a pocket-sized battery won't require the same power to charge then over 600 average suburban homes. 

## Changelog
:cl:
balance: suit and cell chargers should draw from grid preventing instant apc blackouts in most cases.
/:cl:

### CODE REVIEW

**Review by 00-Steven — COMMENTED** [review]
This should have a changelog as it's a player-facing change.

**Comment by Blacklist897** [comment]
it seems i have accidently cut the changelog off

**Review by 00-Steven — CHANGES_REQUESTED** [review]
Annoyingly I don't seem to be able to comment on parts of the code you didn't touch, but there's a few more places you've missed.
https://github.com/tgstation/tgstation/blob/9f37c8376255e6906d22c95d246c174a5d56b656/code/game/machinery/cell_charger.dm#L10
With your changes, this should be 25 and not 250.
https://github.com/tgstation/tgstation/blob/9f37c8376255e6906d22c95d246c174a5d56b656/code/game/machinery/suit_storage_unit.dm#L290
And this should be 5 and not 50.

**Comment by Blacklist897** [comment]
ok I fucked up and made it a draft AAAAAAAAAAAAAAA

**Review by 00-Steven — APPROVED** [review]
Won't comment much on the balance side, but code-wise I think that's all the related values accounted for now.

**Comment by Blacklist897** [comment]
this is fine to undraft btw

**Comment by MrMelbert** [comment]
For the record, you can undraft it yourself with this "Ready for Review" button. 

image

**Comment by SyncIt21** [comment]
This also means that cells charge slower and I don't see how that's benificial.

If people's concers are cell chargers blacking out the room when there isn't enough power then there is a very simple fix.

Just make sure you use only power from the grid and not the  apc cell over here
https://github.com/tgstation/tgstation/blob/9f37c8376255e6906d22c95d246c174a5d56b656/code/game/machinery/cell_charger.dm#L143-L144

There is a 2nd param you can pass to both these procs. 
```
use_energy(main_draw * 0.01, grid_only = TRUE) //use a small bit for the charger itself, but power usage scales up with the part tier 
charge_cell(main_draw, charging, grid_only = TRUE) 
```

I did this for mod suit rechargers.

**Comment by Pickle-Coding** [comment]
This would take 400 seconds to recharge a cell of the same tier. That's a fairly harsh reduction.

Lower it to like 150kW, then the roundstart charger will take about 68 seconds to charge a cell of the same tier, while leaving some room from the grid after engineers set the SMES.

**Comment by Pickle-Coding** [comment]
Maybe in the future I could make an argument that prevents it drawing energy from the APC below a certain percentage, so we can speed up charging with APC energy without also blacking out the room.

**Comment by zxaber** [comment]
Cell chargers are supposed to pull from the powernet directly when charging a cell. We went through this same thing like two years ago and I changed it so that they'd impact the SMES units instead.

**Comment by Blacklist897** [comment]
Im going to update the values to around pickles ones 
but I don't really understand how to make them draw from smes

**Comment by SyncIt21** [comment]
By smes they mean grid. Just put back the original values and do this instead https://github.com/tgstation/tgstation/pull/82309#issuecomment-2027898991
Find the same procs for suit storage unit, put back the original values and do this

**Comment by FlufflesTheDog** [comment]
I could be wrong, but I'm pretty sure they DO pull from the grid.
The problem is that they pull an entire stations worth of power; 250kw-1mw depending on upgrades, whereas the supermatter typically generates 1-1.2mw, and before the rework a minimal use station would draw around 300kw. And the absolute most 3 unupgraded smes units will put out is 600kw. Thus they instantly overtax the grid and fall back on using the apc cell like any other device does, which of course instantly drains it. Making it refuse to draw from the apc directly would help but sort of hides the problem, in my opinion.

This is sort of the consequence of energy being conserved, without having an appropriate distinction between space station sized batteries and  microwave sized batteries. The entire station has a whopping 150MJ of high capacity cells to store 10-20 minutes of station power, and someone charging 4 bluespace cells is enough to completely drain that.

**Comment by SyncIt21** [comment]
That's the thing we don't want to touch the apc cell evem when there isn't enough power in the grid, So it becomes fine to charge cells without causing a blackout.

Nothing can be done about the power requirements but we can atleast save the room which takes higher priority

**Comment by Blacklist897** [comment]
yeah the issue is chargers drawing a megawatt each and this will slow that process down.
as a side effect it also makes using a wrenched cell charger as a powersink to kill ai a lot slower

**Comment by Iajret** [comment]
> I could be wrong, but I'm pretty sure they DO pull from the grid. The problem is that they pull an entire stations worth of power; 250kw-1mw depending on upgrades, whereas the supermatter typically generates 1-1.2mw, and before the rework a minimal use station would draw around 300kw. And the absolute most 3 unupgraded smes units will put out is 600kw. Thus they instantly overtax the grid and fall back on using the apc cell like any other device does, which of course instantly drains it. Making it refuse to draw from the apc directly would help but sort of hides the problem, in my opinion.
> 
> This is sort of the consequence of energy being conserved, without having an appropriate distinction between space station sized batteries and microwave sized batteries. The entire station has a whopping 150MJ of high capacity cells to store 10-20 minutes of station power, and someone charging 4 bluespace cells is enough to completely drain that.

Some balancing should be done here, indeed. Either making smeses magically store 1000x energy of regular cells they are made of (by making said cell store kilojoules instead of megajoules) or adding completely new item for them? Latter is worse indeed. We don't need new part that serve no purpose outside of single thing (im looking at you, telecomm machinery). But then, you could not build precharged SMESes for better or worse.

tldr: i think cells of every tier should have their capacity reduced by 10, 100 or even 1000 times while leaving SMESes capacity untouched.

**Comment by zxaber** [comment]
Changing cell capacity is tedious because you have to touch every item that uses cells too. It's probably easier to multiply SMES storage and SM output instead.

For what it's worth, if we did ever get super-sized cells, I'd probably change mechs to use those rather than normal cells. So they wouldn't be limited *solely* to SMES units.

(I think crafting 3 cells with an iron rod to make a "cell bank" would be neat and an easy way of implementing large cells.)

**Comment by SyncIt21** [comment]
> (I think crafting 3 cells with an iron rod to make a "cell bank" would be neat and an easy way of implementing large cells.)

Why don't we just add more cells to the smes required component parts? Make it from 5 cells to 15 or something, that way we can scale capacity without magic number multiplication.

**Comment by lessthnthree** [comment]
> > It's probably easier to multiply SMES storage and SM output instead.
> 
> Why don't we just add more cells to the smes required component parts? Make it from 5 cells to 15 or something, that way we can scale capacity without magic number multiplication.
> 
> If we are multiplying SM output then we need to talk about scaling a lot of other power sources as well(turbines, solars, plasma generators etc) which is a headache. Their energy is sent to the smes anyway so even though they take longer to charge now unlike before it's not a big deal after the smes get's fully charged and takes over.

Yeah, like how a power storage unit is dozens of batteries in reality, why can't an SMES take a bunch more. The resource cost of upgrading cells is already minimal.

**Comment by Pickle-Coding** [comment]
APC cells are upgraded ones, which is 2.5 MJ (worse than the high capacity ones which everyone uses at minimum, which is 10MJ).

It's worth noting that the combined APC cells in their current state hold a large composition of the roundstart energy supply, more than the SMES, so upgrading SMES cell would have a significant impact on the time it takes for power to run out (which is around 20 minutes).

I think one solution is to like, not standardise high capacity cells as the norm. They should be larger than normal cells, and only fit in certain machinery like SMES and mechs. The problem is that a lot of things are balanced around the high capacity cell scale, so we would have to scale everything again, which will take a lot of effort.

**Comment by Pickle-Coding** [comment]
Also for what it's worth, the APC charge rate is capped at 1% per second. 25kW will charge the standard APC in 100 seconds. Is that considered long for you, or is it longer because of lag?

**Comment by lessthnthree** [comment]
> Also for what it's worth, the APC charge rate is capped at 1% per second. 25kW will charge the standard APC in 100 seconds. Is that considered long for you, or is it longer because of lag?

The APC charge time has no issues. I was thinking in the context of restricting cell chargers of an area being unable to exceed the charge rate of the area APC, so that they don't drain the APC faster than it can charge. 25kW is 1/10 of the current tier 1 cell charger rate, let alone tier 4.

**Comment by lessthnthree** [comment]
> APC cells are upgraded ones, which is 2.5 MJ (worse than the high capacity ones which everyone uses at minimum, which is 10MJ).
> 
> It's worth noting that the combined APC cells in their current state hold a large composition of the roundstart energy supply, more than the SMES, so upgrading SMES cell would have a significant impact on the time it takes for power to run out (which is around 20 minutes).
> 
> I think one solution is to like, not standardise high capacity cells as the norm. They should be larger than normal cells, and only fit in certain machinery like SMES and mechs. The problem is that a lot of things are balanced around the high capacity cell scale, so we would have to scale everything again, which will take a lot of effort.

What if room APCs were hypers but their initial charge was the same 2.5MJ they have currently? This would keep the same amount of roundstart power in the system, however leave more space for APCs to charge once the engine is running. 

The potential side effect however is that areas would be able to survive longer with a loss of power, assuming their APCs had time to charge.

**Comment by CRITAWAKETS** [comment]
> Some balancing should be done here, indeed. Either making smeses magically store 1000x energy of regular cells they are made of (by making said cell store kilojoules instead of megajoules) or adding completely new item for them? Latter is worse indeed. We don't need new part that serve no purpose outside of single thing (im looking at you, telecomm machinery). But then, you could not build precharged SMESes for better or worse.

We could also just lean into realism and make Superconducting Magnetic Energy Storage work realistically, it's not meant to be a battery bank but instead a superconducting wire coil that holds magnetic charge. It wouldn't require cells, just some cable coil and presumably a laser to cool the system down. That said, this does come with the issue of no pre-charged SMES cells, but those are kind of a bad idea when it comes to balance anyway when it comes to thinks like potato cells.

**Comment by Pickle-Coding** [comment]
> > Also for what it's worth, the APC charge rate is capped at 1% per second. 25kW will charge the standard APC in 100 seconds. Is that considered long for you, or is it longer because of lag?
> 
> 
> 
> The APC charge time has no issues. I was thinking in the context of restricting cell chargers of an area being unable to exceed the charge rate of the area APC, so that they don't drain the APC faster than it can charge. 25kW is 1/10 of the current tier 1 cell charger rate, let alone tier 4.

If it's draining the APC, then the APC isn't charging, because cell chargers and the vast majority of machines prioritise the grid first. The grid typically doesn't have enough power to charge the cell at 250kW. It would be better to just make cell chargers ignore the APC cell, because 25kW is 1,000 seconds for a high capacity cell.

I'm not sure how many people know about this, but SMES only discharge the energy that actually gets used. We could upgrade tier 1 SMES IO to 250kW each, then there would be enough excess power to at least charge cells with tier 1 chargers, at least temporarily even if the engine can't sustain that indefinitely.

**Comment by lessthnthree** [comment]
> > > Also for what it's worth, the APC charge rate is capped at 1% per second. 25kW will charge the standard APC in 100 seconds. Is that considered long for you, or is it longer because of lag?
> > 
> > 
> > The APC charge time has no issues. I was thinking in the context of restricting cell chargers of an area being unable to exceed the charge rate of the area APC, so that they don't drain the APC faster than it can charge. 25kW is 1/10 of the current tier 1 cell charger rate, let alone tier 4.
> 
> If it's draining the APC, then the APC isn't charging, because cell chargers and the vast majority of machines prioritise the grid first. The grid typically doesn't have enough power to charge the cell at 250kW. It would be better to just make cell chargers ignore the APC cell, because 25kW is 1,000 seconds for a high capacity cell.

Yeah, changing the cell chargers back to be grid_only = TRUE is likely the only avenue without a bunch of pain until the overall power consumption ecosystem is looked at.

**Comment by Iajret** [comment]
> We could also just lean into realism and make Superconducting Magnetic Energy Storage work realistically, it's not meant to be a battery bank but instead a superconducting wire coil that holds magnetic charge. It wouldn't require cells, just some cable coil and presumably a laser to cool the system down. That said, this does come with the issue of no pre-charged SMES cells, but those are kind of a bad idea when it comes to balance anyway when it comes to thinks like potato cells.

Thats actually sounds kinda good. But still leaves the main problem untouched. T4 charger takes 1 MW and 40 secs to charge T4 cell. Entire station has a load of 300-600 (maybe 800 kW when botany or atmospherics decides to go insane). You need to have 1 MW available for each charger. It was already have been said that lowering maximum power for chargers will just make charging more of a chore (although, it turned out, you do not need to charge a lot of cells) because it will take a lot longer to charge something. You go from 1 MW to 0.5 MW. It now takes 80 secs and still consumes a fucking lot of power. And leaving surplus of literal megawatts in the net makes hacking doors or touching random grilles just outright deadly.

**Comment by SyncIt21** [comment]
> It would be better to just make cell chargers ignore the APC cell

That's what I've been suggesting this whole time. Thank you for reconfirming that 

> Yeah, changing the cell chargers back to be grid_only = TRUE

Exactly this is how. @Blacklist897 put back the original values and follow the suggestion I made in https://github.com/tgstation/tgstation/pull/82309#issuecomment-2027898991 find the relevant code for suit storage units and do the same

**Comment by zxaber** [comment]
> can someone else weigh in on if that's a good idea please

See #54989. What Synclt is suggesting is something I had tried to accomplish (and to my knowledge, was working fine up until the recent power changes), except I'd only taken the scope of cell chargers alone.

**Comment by Blacklist897** [comment]
Since I lack the ability to unfuck a single cell charger drawing the same power as 650 homes (as power values are wild) im going to go with SyncIt21's idea and make em draw from grid

**Comment by lessthnthree** [comment]
Remember to change the title/changelog that cell chargers are now bypassing APC, instead of reduced power.
